### ISSUE TITLE
Convert all core JS to Typescript

### ISSUE THREAD

**Issue by askvortsov1** [flarum/core#2342]
This should be done in small, maximally isolated PRs: no more than a few files updated per PR so that we can slowly review and phase these in over time.

**Comment by mihirgupta0900** [comment]
Hey maintainers! 
I would love to take this issue up, looks like an amazing learning opportunity! I'm not very experienced with Typescript, but know the basics.

**Comment by askvortsov1** [comment]
That's great to hear, and we'd definitely appreciate the help! https://docs.flarum.org/contributing.html might be a good place to start for setup stuff (and reading a bit more about how flarum works). Our [community forums](http://discuss.flarum.org/) and [discord chat](https://flarum.org/discord) are also good resources if you have questions / run into issues.

**Comment by mihirgupta0900** [comment]
Great to hear! I've joined the discord chat and I'll go through the links in a few hours and get started soon!

**Comment by knoxd8256** [comment]
Is this a good issue for multiple contributors? I'd also like to work on this, and I've joined the discord so we might make sure not to double up on work.

**Comment by tankerkiller125** [comment]
@knoxd8256 this would be a great issue for multiple contributors, my recommendation would be to negotiate work being done via our discord chat and/or this issue so that we don't have repeat file updates.

**Comment by nina-py** [comment]
Can I confirm that the file naming conventions for TypeScript files are `.ts` for general code and `.tsx` for anything containing JSX-like syntax? The linked PRs seem to be following these naming conventions but the unconverted JS files are all `.js`. The reason I'm asking is that if `.tsx` files are the new norm for components and the like, this line in `tsconfig.json` needs to be updated to include them to get rid of certain warnings, i.e. the `esModuleInterop` one:

https://github.com/flarum/core/blob/9cb9097b24e0345816f2f8ef22789ab804e796f6/js/tsconfig.json#L2

**Comment by datitisev** [comment]
@nina-py Yes, we use `.tsx` for JSX.

> this line in `tsconfig.json` needs to be updated to include them to get rid of certain warnings, i.e. the `esModuleInterop` one:

Hm, if that's the source of the warning then that's a big 🤦 for me. Feel free to PR to that - it might be enough if you just add `*` after `*.ts`, can't remember if that works or not.

**Comment by askvortsov1** [comment]
Maybe that might also help with the ESDoc generation issue? Although I doubt it

**Comment by datitisev** [comment]
@askvortsov1 Don't think so - that's a separate config fiel anyway.

**Comment by nina-py** [comment]
Thanks @datitisev, a PR is on the way.

> Hm, if that's the source of the warning then that's a big facepalm for me. Feel free to PR to that - it might be enough if you just add * after *.ts, can't remember if that works or not.

According to the docs (see quote below), all that needs to be done is removing the `.ts.` file extension and it will automatically pick up only the converted files and not the entire codebase. 

> If a glob pattern doesn’t include a file extension, then only files with supported extensions are included (e.g. .ts, .tsx, and .d.ts by default, with .js and .jsx if allowJs is set to true). 

https://www.typescriptlang.org/tsconfig#include

**Comment by nina-py** [comment]
Also, there are more warnings in the files that's already been converted to TypeScript. Wouldn't it be great if these were caught automatically by a GitHub action? ESLint has a plugin for TypeScript, perhaps it could be added to the workflow?

**Comment by datitisev** [comment]
@nina-py Resolving TS warnings is not a priority because we're using TS to help typehinting mainly, not to have perfect code (in terms of TS warnings). If you want to PR fixes for warnings tho, go ahead (unless they make the code ugly, then we may ignore it).

**Comment by nina-py** [comment]
Thanks @datitisev, I certainly don't have an intention of making any code ugly. Either way, I need to read up on Mithril and get a bit more comfortable with the Flarum frontend code before starting on this.
